### PR TITLE
[FLINK-33986][runtime] Extend ShuffleMaster to support snapshot and restore state.

### PR DESCRIPTION
**Pull request by JunRuiLee**
## What is the purpose of the change

Extend ShuffleMaster to support snapshot and restore state.

## Brief change log

Extend shuffleMaster to support batch snapshot as follows:

1. Add method supportsBatchSnapshot to identify whether the shuffle master supports taking snapshot in batch scenarios
2. Add method snapshotState and restoreState to snapshot and restore the shuffle master's state.

## Verifying this change

This change is a trivial rework / code cleanup without any test coverage.

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (yes / **no**)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (yes / **no**)
  - The serializers: (yes / **no** / don't know)
  - The runtime per-record code paths (performance sensitive): (yes / **no** / don't know)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn, ZooKeeper: (yes / **no** / don't know)
  - The S3 file system connector: (yes / **no** / don't know)

## Documentation

  - Does this pull request introduce a new feature? (yes / **no**)
  - If yes, how is the feature documented? (**not applicable** / docs / JavaDocs / not documented)

### CODE REVIEW

File: flink-runtime/src/main/java/org/apache/flink/runtime/shuffle/ShuffleMasterSnapshotContext.java [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.runtime.shuffle;
+
+/**
+ * Snapshot context used to create {@link ShuffleMasterSnapshot}, which can provide necessary
+ * information.
```
Comment (zhuzhurk) [inline flink-runtime/src/main/java/org/apache/flink/runtime/shuffle/ShuffleMasterSnapshotContext.java:23]: > which can provide necessary information

Looks redundant to me. I prefer to remove it.

File: flink-runtime/src/main/java/org/apache/flink/runtime/shuffle/ShuffleMaster.java [hunk 2]
```diff
@@ -121,4 +122,21 @@ default CompletableFuture<Collection<PartitionWithMetrics>> getAllPartitionWithM
             JobID jobId) {
         return CompletableFuture.completedFuture(Collections.emptyList());
     }
+
+    /**
+     * Whether the shuffle master supports taking snapshot in batch scenarios, which will be used
+     * when enable Job Recovery. If it returns true, we will call {@link #snapshotState} to take
```
Comment (zhuzhurk) [inline flink-runtime/src/main/java/org/apache/flink/runtime/shuffle/ShuffleMaster.java:128]: which will be used when enable Job Recovery
-> if `execution.batch.job-recovery.enabled` is true.

we
-> Flink

**Comment by JunRuiLee** [comment]
Thanks @zhuzhurk for review, I've updated this pr accordingly. PTAL~

**Review by zhuzhurk — APPROVED** [review]
LGTM.

**Pull request merged**